### PR TITLE
add GenerationChanged predicate filter to all resources

### PR DESCRIPTION
**Pull request by cnvergence**
**What type of PR is this?**

**What this PR does / why we need it**:
Per review of https://github.com/envoyproxy/gateway/pull/1938#discussion_r1358956623, it would be valuable to add GenerationChanged predicate to all resources.
This skip events that have no change in the object Spec and only the metadata and/or status fields are changed.

### CODE REVIEW

File: internal/provider/kubernetes/controller.go [hunk 1]
```diff
@@ -1296,13 +1296,15 @@ func (r *gatewayAPIReconciler) watchResources(ctx context.Context, mgr manager.M
...
 		predicate.NewPredicateFuncs(r.hasMatchingController),
+		predicate.GenerationChangedPredicate{},
 	); err != nil {
 		return err
 	}
 
 	// Only enqueue EnvoyProxy objects that match this Envoy Gateway's GatewayClass.
 	epPredicates := []predicate.Predicate{
 		predicate.ResourceVersionChangedPredicate{},
+		predicate.GenerationChangedPredicate{},
```
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:1307]: is the predicate func as OR op or an AND op ?
if its an AND op then we should be adding `GenerationChangedPredicate` to the beginning of the list

**Review by arkodg — APPROVED** [review]
LGTM thanks !
should hopefully reduce reconciles

**Review by zhaohuabing — APPROVED** [review]
LGTM. Thanks for the optimization!

**Review by zirain — APPROVED** [review]
Thanks!

**Comment by zirain** [comment]
@cnvergence sorry, I merge this by mistake, can you resubmit this and fix the CI?

**Pull request merged**